public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx

```diff
@@ -374,10 +263,8 @@ export class ConfigEditor extends PureComponent<Props, State> {
               <Input
                 className="width-30"
                 placeholder="Namespace1,Namespace2"
-                value={config.jsonData.customMetricsNamespaces || ''}
-                onChange={(event: ChangeEvent<HTMLInputElement>) =>
-                  this.onCustomMetricsNamespacesChange(event.target.value)
-                }
+                value={options.jsonData.customMetricsNamespaces || ''}
+                onChange={onUpdateDatasourceJsonDataOption(this.props, 'customMetricsNamespaces', false)}
```

Comment:
The only downside to this approach is that we have strings rather than typed properties as the key we set.

@dprokop, @aocenas -- do you have any ideas how to make it based on a typescript property? (even if it has to be entered as a string)

Reply:
I think what you could do is to slightly refactor the onUpdateDatasourceJsonDataOption function and make it generic:

```ts
export const onUpdateDatasourceJsonDataOption<J,S, K extends keyof S> = (
  props: DataSourcePluginOptionsEditorProps<J,S>,
  key: K,
  secure: boolean
) => (event: React.SyntheticEvent<HTMLInputElement | HTMLSelectElement>) => {
  updateDatasourcePluginJsonDataOption(props, key, event.currentTarget.value, secure);
};
```

I'm not sure the snippet above works, but this is the direction to get the typed keys

Reply:
As @dprokop said using something like `keyof Props` will give you type that accepts only string that match the keys of the Prop type (and it does not need to be generic if it is not needed for something else). I used something like that [here](https://github.com/grafana/grafana/blob/fc251816eb900b0a1c16929aae907d8b38fed7c9/public/app/plugins/datasource/loki/configuration/ConfigEditor.tsx#L10) for example and it type checks the string `field` arg fine.